### PR TITLE
Block registration: normalize blockType.parent to `array`

### PR DESCRIPTION
**Pull request by oandregal**
Follow-up to https://github.com/WordPress/gutenberg/pull/66234

## What?

Casts `blockType.parent` to array if it's a string.

## Why?

While we haven't enforced it in any way, we have been operating under the assumption that `blockType.parent` is an array. Apparently, not always is and blocks may register it using a string. When processing input from 3rd parties, we should enforce the type so it's safe to operate with it during the whole lifecycle of the block (e.g.: to avoid issues like [this one](https://github.com/WordPress/gutenberg/pull/66234)).

## How?

- Removes the hotfix introduced at https://github.com/WordPress/gutenberg/pull/66234
- Cast to array in block registration 97cb5b53dc314651ef36dfe73c42cbb8cd37633e
- Cast to array in `'blocks.registerBlockType'` filter fe4e3bece5e5375a2ca6d5432e25ad6a3379f4f9

## Testing Instructions

Update the `parent` of a core block to be a string instead of array (e.g.: update the column block).

Test the following scenarios:

- Inserter:
  - Child block is listed but cannot be added if there's no parent.
  - If the parent block type is selected, the child block shows up as "suggested" and clicking on it adds a new child.
- In place inserter (`+` button):
  - Child block is not listed.
  - The parent block has specific affordances to add new children.

Trunk

- It's possible to add a child block (column) as top-level block. It shouldn't.
- It's possible to add a child block (column) within the same block type (column). It shouldn't.

https://github.com/user-attachments/assets/5c2a24a2-cb4c-4144-a956-37db8a380673

This PR

https://github.com/user-attachments/assets/5ecef613-c1f4-4924-8d78-7a36aa92a55f

### CODE REVIEW

**Comment by oandregal** [comment]
This is ready to review.

I can confirm that using a string as a parent worked (registered the block), though it had side-effects (child blocks could be added top-level + could be added within themselves). This PR fixes that and normalises `blockType.parent` to an array, so it's safe to operate with it during the whole block lifecycle.

Added some testing instructions and covered both block registration + the `blocks.registerBlockType` filter. Is there any other place I should check?

File: packages/blocks/src/api/registration.js
Comment (oandregal) [inline packages/blocks/src/api/registration.js]: Even if we migrate strings to array, I think we should log a warning to educate block authors. Thoughts?

File: packages/blocks/CHANGELOG.md
Comment (oandregal) [inline packages/blocks/CHANGELOG.md]: Added a changelog to make sure this doesn't go unnoticed.

File: packages/blocks/src/api/registration.js
Comment (oandregal) [reply packages/blocks/src/api/registration.js]: Actually, I had to remove it at https://github.com/WordPress/gutenberg/pull/66250/commits/32c9a1f62805b1c46b3d3e928810cc0181e26aa4 for tests to pass.

File: packages/blocks/CHANGELOG.md
Comment (ntsekouras) [reply packages/blocks/CHANGELOG.md]: We should also add a heading to the change log entry, but I'm not sure what that should be.. Maybe `code quality`? 🤔

Comment (oandregal) [reply packages/blocks/CHANGELOG.md]: Added `Bug fixes` heading https://github.com/WordPress/gutenberg/pull/66250/commits/0c4a9f55ec3e9f081274255691b27a79757c939d

File: packages/blocks/src/api/registration.js
Comment (ntsekouras) [inline packages/blocks/src/api/registration.js]: Isn't the `parent` prop in metadata (if an object is passed)?  If that's the case and the metadata are not passed we should get them and make the check.

Additionally can we simplify the checks here to explicitly check what we want? Something like:

```
if (
	blockNameOrMetadata.parent &&
	! Array.isArray( blockNameOrMetadata.parent ) &&
	typeof blockNameOrMetadata.parent === 'string'
) {
	blockNameOrMetadata.parent = [ blockNameOrMetadata.parent ];
}
```

File: packages/blocks/src/api/registration.js
Comment (mcsf) [reply packages/blocks/src/api/registration.js]: Can we amend the tests to expect the warnings? Seems like a better outcome.

File: packages/block-editor/src/store/selectors.js
Comment (mcsf) [inline packages/block-editor/src/store/selectors.js:1632]: It's a matter of preference, of course, but duck-typing can be less obtrusive here:

```suggestion
	return blockType.parent.some?.(
```

File: packages/blocks/CHANGELOG.md
Comment (mcsf) [inline packages/blocks/CHANGELOG.md]:
```suggestion
- Enforce `blockType.parent` to be an array. While string values were never supported, they appeared to work with some unintended side-effects that have been fixed by [#66250](https://github.com/WordPress/gutenberg/pull/66250). For that reason, we've added some code that automatically migrates strings to arrays — though it still raises a warning.
```

Other notes:
* **enforce:** I'd either go with "cast" if we decide to do it automatically with no warnings, or "require" if we add warnings; enforce is more ambiguous to me.

**Review by mcsf — APPROVED** [review]
Looks good, but left notes

File: packages/blocks/src/api/registration.js
Comment (oandregal) [reply packages/blocks/src/api/registration.js]: > Isn't the parent prop in metadata (if an object is passed)? If that's the case and the metadata are not passed we should get them and make the check.

Oh, I missed that, thanks for bringing it up! I'm a bit confused about this for some reasons (existing code, [docs](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-registration/) and other [implementations](https://github.com/WordPress/gutenberg/blame/update/cast-string-parent-to-array/packages/block-library/src/index.native.js#L136) tell otherwise), but specially because the PR worked in my tests.

> Additionally can we simplify the checks here to explicitly check what we want?

What we want to do is "if the parent is a string, normalize it to an array". In my view, the current code is actually more explicit, simpler (why to check both "is it not an array" and "it is a literal string?"), and more comprehensive (String objects).

Comment (oandregal) [reply packages/blocks/src/api/registration.js]: So, we actually have to use settings, just not in `registerBlockType` but in `processBlockType`. This is the data flow for block registration:

- `registerBlockType( nameOrMetadata, settings )`
  - `addBootstrappedBlockType` => if metadata provided, creates a new tree in the store with that data
  - `addUnprocessedBlockType`
    - calls `processBlockType` => takes the existing block metadata & merges it with the settings
      - **merges metadata & settings** into `blockType`
      - dispatch `'blocks.registerBlockType'` filter with `blockType`
      - validates the `blockType` after the filter has been dispatched
    - creates a new tree in the store with the blockType data
  - return `blockType` (settings & metadata consolidated)

I've moved it all to `processBlockType` (including existing validation for the parent that was done too early) in [002ac7cec3337f5f5d56c9f154a3e337b890f3c4](https://github.com/WordPress/gutenberg/pull/66250/commits/002ac7cec3337f5f5d56c9f154a3e337b890f3c4).

Comment (oandregal) [reply packages/blocks/src/api/registration.js]: After having looked at this more deeply, there's a situation in which this could be considered a breaking change:

- 3rd party code registers the blockType with parent as a string.
  - Before: parent will be a string in the store.
  - After: parent will be an array in the store.

Note that I intentionally added the normalization code _after_ the filter, as to not interfere with 3rd party code before they executed their callback. However, if the 3rd party code queries the store for the parent _after registration_ to do something with the `parent`, it will no longer be a string.

I don't see a way out of this and it seems to me that this PR fixes an important issue (being able to add a child block within itself). This PR will also be merged very early in the WordPress 6.8 cycle, so we'll have time to gather feedback.

Thoughts?

File: packages/blocks/src/api/registration.js
Comment (oandregal) [reply packages/blocks/src/api/registration.js]: Yeah, this is better. I was also reluctant to add a warning but given this can have [side-effects](https://github.com/WordPress/gutenberg/pull/66250#discussion_r1810568607), it's best to have it.

File: packages/block-editor/src/store/selectors.js
Comment (oandregal) [reply packages/block-editor/src/store/selectors.js:1632]: I'd rather be explicit (check for array).

**Comment by oandregal** [comment]
I'd welcome a double check of https://github.com/WordPress/gutenberg/pull/66250#discussion_r1810559780 before merging this PR — I'll let it sit for a few days and will merge if no one opposes.

cc @gziolo @youknowriad @ntsekouras @mcsf

File: packages/blocks/src/store/process-block-type.js
Comment (ntsekouras) [inline packages/blocks/src/store/process-block-type.js:201]: Copied the comment from the other PR:

> > I'm curious why we'd need both typeof blockType.parent === 'string' || blockType.parent instanceof String.. 

> Otherwise, we don't cover all potential issues. Note the issue was that we checked for `.length` in a type that wasn't an array (didn't have `.some`). That's the case for both literal & object strings.

I still don't see why is needed. Can you give me an example?

In general my [suggestion before ](https://github.com/WordPress/gutenberg/pull/66250#discussion_r1810049802)for simplifying the checks by what we expect and not what we don't expect. With that code `parent` could be a boolean etc.. 

And if we wanted to be thorough we should also check if an array is provided, that consists of `strings`.

File: packages/blocks/src/api/registration.js
Comment (ntsekouras) [reply packages/blocks/src/api/registration.js]: >which this could be considered a breaking change

That's true, but I'm not sure if we should consider a breaking change something that was used in a wrong way and differently than documentation. We had a similar issue with `description` in the past, where some plugins would provide a component and we handled that the same way as you did (with a small difference `deprecated` vs `warning`).

Additionally, it sounds unusual to use `parent` from store and this PR seems good to me regarding this aspect.

File: packages/blocks/src/store/process-block-type.js
Comment (mcsf) [inline packages/blocks/src/store/process-block-type.js:227]: I understand this check was carried from before, but why the check for `1`? Wouldn't the following be better in all respects?

```suggestion
		if ( settings?.parent?.includes?.( name ) ) {
```

**Review by mcsf — COMMENTED** [review]
I'm glad we caught this and are targeting `processBlockType` now.

File: packages/blocks/src/api/registration.js
Comment (mcsf) [reply packages/blocks/src/api/registration.js]: As long as we log the offence, I'm ok with this. Any third-party logic interested in monitoring blocks' `parent` setting would quickly notice that core blocks use arrays, unless that logic were really built very specifically for a third-party block using `parent` as a string. It's possible that someone is doing that, but not too likely.

We _can_ play it safe and communicate this as a breaking change in dev notes, even though it isn't.

File: packages/blocks/src/store/process-block-type.js
Comment (oandregal) [reply packages/blocks/src/store/process-block-type.js:201]: Ok, these are the facts:

- I discovered that we were trying to detect arrays the wrong way while responding to [Gutenberg forum issues](https://wordpress.org/support/topic/error-after-update-228/). We [were checking](https://github.com/WordPress/gutenberg/pull/65700#discussion_r1806315983) for `blockType.parent?.length`, but that's not enough to detect arrays — it can be a string, any type of it, as it was. I prepared a [hotfix](https://github.com/WordPress/gutenberg/pull/66234) but wanted to tackle the core issue: that we didn't normalize the parent data coming from 3rd parties — that's what this PR is for.  I presume we have more of these in our codebase are causing issues for the same reason.

- Strings can be [literal or object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String#string_primitives_and_string_objects). Both have the `.length` property and are [boxed/unboxed](https://oandre.gal/articles/values-and-references-in-javascript/#values-references) depending on what you do. So 

```js
console.log( 'string'.length ); // Logs 6
console.log( new String( 'string' ).length ); // Logs 6
```

log the same.

Back to our problem. Block authors should have been using arrays to declare the parent, but, in some cases, they weren't — instead they used strings of any type. They can use either type, that's why we need to cover for both.

Does this help?

File: packages/blocks/src/store/process-block-type.js
Comment (oandregal) [reply packages/blocks/src/store/process-block-type.js:227]: I'm not sure and it increases the surface area of this PR that we need to test, etc. — I'd rather do that separately. For reference, this is the [issue](https://github.com/WordPress/gutenberg/issues/66090) and [PR](https://github.com/WordPress/gutenberg/pull/66121).

**Comment by oandregal** [comment]
> **This would be the DevNote for this change (see [conversation](https://github.com/WordPress/gutenberg/pull/66250#discussion_r1810568607)).** (I'll keep this updated if there are changes/suggestions/etc.).

For blocks registered using a string for the `parent` property, we now automatically migrate to the proper type, an `array`.

While string values were never officially supported for the `parent` property, they appeared to work with some unintended side-effects that have been fixed by [#66250](https://github.com/WordPress/gutenberg/pull/66250). For that reason, we've added some code that automatically migrates strings to arrays.

- If all you did with the `parent` property was registering the block, your code will continue to work as before — though a warning will be issued and it's recommended that you migrate your blocks to the proper syntax.
- If your code was querying the Redux store and operating with the `parent` property of blocks, we recommend you review it to make sure it works well with arrays. In principle, it should work as before given that's the proper type for `parent`. However, your code may need to be updated if you were doing something like:

```js
const blockType = select( blocksStore ).getBlockType( name );
if ( name === 'my/block' ) {
  // Operate with blockType.parent as if it was a string.
}
```

File: packages/blocks/src/api/registration.js
Comment (oandregal) [reply packages/blocks/src/api/registration.js]: > We can play it safe and communicate this as a breaking change in dev notes, even though it isn't.

Added the label and [some text](https://github.com/WordPress/gutenberg/pull/66250#issuecomment-2429493919) for the devnote.

File: packages/blocks/src/store/process-block-type.js
Comment (mcsf) [reply packages/blocks/src/store/process-block-type.js:227]: Valid point!

**Comment by oandregal** [comment]
Merged this one. Happy to continue conversation or address any follow-ups people run into.

**Pull request merged**
